File: task2.py
import sys

arr = [int(value) for value in ' '.join(sys.stdin.readlines()).strip().split(' ')]
arr_sort = []

for i in range(0, len(arr)-1):
    arr_sort.append([arr[i+1]-arr[i], i, arr[i+1], arr[i]])

arr_sort.sort(key=lambda value: value[0])
all_count = sum(arr)

count = 0
for i in range(len(arr_sort)-1,0,-1):
    value = arr_sort[i]
    if arr[value[1]] != 0 and arr[value[1]+1] !=0:
        tmp_count = count + arr[value[1]+1]*2
        tmp_all_count = all_count - arr[value[1]+1] - arr[value[1]]

        if (tmp_count+tmp_all_count<count+all_count):
            break

        arr[value[1]] = 0
        arr[value[1]+1] = 0
        count = tmp_count
        all_count = tmp_all_count

print(count+all_count)

File: task2.rb
class Pair
  attr_reader :diff

  def initialize(index, score)
    @score = score
    @index = index
    @diff = @score[@index+1]-@score[@index]
  end

  def <=> value
    @diff <=> value.diff
  end

  def bonus
    @score[@index+1]*2
  end

  def exclude!
    @score.fill 0, @index, 2
  end

  def excluded?
    @score[@index]==0 || @score[@index+1]==0
  end

end

score = readlines.join(' ').split(' ').map { |value| value.to_i }

pairs=[]
for i in 0...score.count-1 do
  pairs.push Pair.new(i, score)
end

pairs.sort!.reverse!

bonus_sum = 0
pairs.each do |pair|
  next if pair.excluded?
  break if pair.diff < 1

  bonus_sum = bonus_sum + pair.bonus
  pair.exclude!
end

puts bonus_sum + score.reduce(:+)

